### PR TITLE
Consolidation of localization

### PR DESCRIPTION
**Pull request by arellegue**

### CODE REVIEW

**Review by David-Engel — APPROVED** [review]

**Review by DavoudEshtehari — CHANGES_REQUESTED** [review]
It seems that time for following the same dynamic generation pattern on .netcore for [ResourceNames](https://github.com/dotnet/SqlClient/blob/a709006541a1169f861ea2c3c6aeb5596cd1c286/src/Microsoft.Data.SqlClient/netcore/src/Resources/StringsHelper.cs#L70) type as well.

Also, I'd rather mix the .netcore-specific "StringsHelper.NetCore.cs" with the new common cs file.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft.Data.SqlClient.csproj [hunk 1]
```diff
@@ -495,9 +495,11 @@
     <Compile Include="..\..\src\Microsoft\Data\SqlClient\SqlStream.cs">
       <Link>Microsoft\Data\SqlClient\SqlStream.cs</Link>
     </Compile>
-    <Compile Include="..\..\src\Resources\ResCategoryAttribute.cs">
+    
+     <Compile Include="..\..\src\Resources\ResCategoryAttribute.cs">
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft.Data.SqlClient.csproj]: The blank line was removed.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft.Data.SqlClient.csproj [hunk 2]
```diff
@@ -495,9 +495,11 @@
...
       <Link>Resources\ResCategoryAttribute.cs</Link>
     </Compile>
+
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft.Data.SqlClient.csproj]: The blank line was removed.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft.Data.SqlClient.csproj [hunk 3]
```diff
@@ -571,19 +573,25 @@
     <Compile Include="Microsoft\Data\SqlClient\SqlConnectionFactory.AssemblyLoadContext.cs" />
   </ItemGroup>
   <ItemGroup Condition="'$(OSGroup)' != 'AnyOS'">
+    
     <Compile Include="Resources\StringsHelper.cs">
       <Link>Resources\StringsHelper.NetCore.cs</Link>
     </Compile>
-    <Compile Include="Resources\Strings.Designer.cs">
+
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft.Data.SqlClient.csproj]: The blank line was removed.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft.Data.SqlClient.csproj [hunk 4]
```diff
@@ -571,19 +573,25 @@
...
+    <Compile Include="..\..\src\Resources\Strings.Designer.cs">
+      <Link>Resources\Strings.Designer.cs</Link>
       <DesignTime>True</DesignTime>
       <AutoGen>True</AutoGen>
       <DependentUpon>Strings.resx</DependentUpon>
     </Compile>
-    <EmbeddedResource Update="Resources\Strings.resx">
+
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft.Data.SqlClient.csproj]: The blank line was removed.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft.Data.SqlClient.csproj [hunk 5]
```diff
@@ -571,19 +573,25 @@
     <Compile Include="Microsoft\Data\SqlClient\SqlConnectionFactory.AssemblyLoadContext.cs" />
   </ItemGroup>
   <ItemGroup Condition="'$(OSGroup)' != 'AnyOS'">
+
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft.Data.SqlClient.csproj]: The blank line was removed.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft.Data.SqlClient.csproj [hunk 6]
```diff
@@ -571,19 +573,25 @@
...
     </Compile>
-    <EmbeddedResource Update="Resources\Strings.resx">
+
+    <EmbeddedResource Include="..\..\src\Resources\Strings.resx">
+      <Link>Resources\Strings.resx</Link>
       <Generator>ResXFileCodeGenerator</Generator>
       <LastGenOutput>Strings.Designer.cs</LastGenOutput>
       <CustomToolNamespace>System</CustomToolNamespace>
     </EmbeddedResource>
+
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft.Data.SqlClient.csproj]: The blank line was removed.

**Comment by arellegue** [comment]
.NetCore ResourceNames auto generation pattern has been added. Also, StringsHelper.NetCore.cs has been merged to StringsHelper.cs.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+{
+    public class LocalizationTest
+    {
+        [Fact]
+        public void Localization_EN_Test()
+        {
+            var localized = "";
+            var expected = "A network-related or instance-specific error occurred while establishing a connection to SQL Server.";
+
+            using TestTdsServer server = TestTdsServer.StartTestServer();
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs:26]: Can you use the `GetLocalizedErrorMessage()` method call?

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+using System.Reflection;
+using System.Security;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.SqlServer.TDS.Servers;
+using Xunit;
+
+namespace Microsoft.Data.SqlClient.Tests
+{
+    public class LocalizationTest
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs:18]: All these tests can be combined if you use `Theory` with `InlineData`.

File: src/Directory.Build.props [hunk 9]
```diff
@@ -43,9 +43,13 @@
...
     <NetCoreSource>$(ManagedSourceCode)netcore\</NetCoreSource>
+	  <NetCoreResources>$(ManagedSourceCode)src\Resources\</NetCoreResources>
+    <ResxFileName>Strings</ResxFileName>
+    <ResourceFileName>SqlClient.Resources.$(ResxFileName)</ResourceFileName>
+    <GeneratedSourceFileName>$(ResxFileName).ResourceNames.cs</GeneratedSourceFileName>
     <NetFxSource>$(ManagedSourceCode)netfx\</NetFxSource>
-    <NetFxResources>$(ManagedSourceCode)netfx\src\Resources\</NetFxResources>
-    <AddOnsPath>$(ManagedSourceCode)add-ons\</AddOnsPath>
+    <NetFxResources>$(ManagedSourceCode)src\Resources\</NetFxResources>
+	    <AddOnsPath>$(ManagedSourceCode)add-ons\</AddOnsPath>
```
Comment (DavoudEshtehari) [inline src/Directory.Build.props:52]: Revert the unnecessary change.

File: src/Directory.Build.props [hunk 10]
```diff
@@ -43,9 +43,13 @@
     <ManagedSourceCode>$(ProjectDir)Microsoft.Data.SqlClient\</ManagedSourceCode>
     <SqlServerSourceCode>$(ProjectDir)Microsoft.SqlServer.Server\</SqlServerSourceCode>
     <NetCoreSource>$(ManagedSourceCode)netcore\</NetCoreSource>
+	  <NetCoreResources>$(ManagedSourceCode)src\Resources\</NetCoreResources>
```
Comment (DavoudEshtehari) [inline src/Directory.Build.props:46]: Indent violation.

File: src/Microsoft.Data.SqlClient/src/Resources/StringsHelper.cs [hunk 11]
```diff
@@ -54,16 +55,69 @@ public static string GetString(string res, params object[] args)
...
+
+        public static string Format(string resourceFormat, object p1, object p2, object p3)
+        {
+            if (UsingResourceKeys())
+            {
+                return string.Join(", ", resourceFormat, p1, p2, p3);
+            }
+
+            return string.Format(resourceFormat, p1, p2, p3);
+        }
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/src/Resources/StringsHelper.cs:121]: netcore specific methods.
```suggestion
#if !NETFRAMEWORK
        // This method is used to decide if we need to append the exception message parameters to the message when calling Strings.Format. 
        // by default it returns false.
        // Native code generators can replace the value this returns based on user input at the time of native code generation.
        // Marked as NoInlining because if this is used in an AoT compiled app that is not compiled into a single file, the user
        // could compile each module with a different setting for this. We want to make sure there's a consistent behavior
        // that doesn't depend on which native module this method got inlined into.
        [MethodImpl(MethodImplOptions.NoInlining)]
        private static bool UsingResourceKeys()
        {
            return false;
        }

        public static string Format(string resourceFormat, params object[] args)
        {
            if (args is not null)
            {
                if (UsingResourceKeys())
                {
                    return resourceFormat + string.Join(", ", args);
                }

                return string.Format(resourceFormat, args);
            }

            return resourceFormat;
        }

        public static string Format(string resourceFormat, object p1)
        {
            if (UsingResourceKeys())
            {
                return string.Join(", ", resourceFormat, p1);
            }

            return string.Format(resourceFormat, p1);
        }

        public static string Format(string resourceFormat, object p1, object p2)
        {
            if (UsingResourceKeys())
            {
                return string.Join(", ", resourceFormat, p1, p2);
            }

            return string.Format(resourceFormat, p1, p2);
        }

        public static string Format(string resourceFormat, object p1, object p2, object p3)
        {
            if (UsingResourceKeys())
            {
                return string.Join(", ", resourceFormat, p1, p2, p3);
            }

            return string.Format(resourceFormat, p1, p2, p3);
        }
#endif
```

File: src/Microsoft.Data.SqlClient/src/tools/scripts/GenerateResourceStringsSource.ps1 [hunk 12]
```diff
@@ -0,0 +1,52 @@
+# Script: GenerateResourceStringsSource.ps1
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/src/tools/scripts/GenerateResourceStringsSource.ps1:1]:
```suggestion
# Licensed to the .NET Foundation under one or more agreements.
# The .NET Foundation licenses this file to you under the MIT license.
# See the LICENSE file in the project root for more information.
# Script: GenerateResourceStringsSource.ps1
```

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+using System;
+using System.Data;
+using System.Data.Common;
+using System.Globalization;
+using System.Reflection;
+using System.Security;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.SqlServer.TDS.Servers;
+using Xunit;
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs:14]: Remove unnecessary usings.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs (hunk 7 above)
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs]: LocalizationTest.cs has been refactored to use [Threory] and [InLineData].

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs (hunk 8 above)
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs:13]: All tests were combined in one test with Theory and InlineData.

File: src/Directory.Build.props (hunk 9 above)
Comment (arellegue) [reply src/Directory.Build.props:52]: Reverted indentation.

File: src/Directory.Build.props (hunk 10 above)
Comment (arellegue) [reply src/Directory.Build.props:46]: Removed indentation. Not sure how it got indented.

File: src/Microsoft.Data.SqlClient/src/Resources/StringsHelper.cs (hunk 11 above)
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Resources/StringsHelper.cs:121]: Enclosed merged code from netcore with #if !NETFRAMEWORK.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs (hunk 13 above)
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs]: Removed unnecessary usings and sorted the rest.

File: src/Microsoft.Data.SqlClient/src/tools/scripts/GenerateResourceStringsSource.ps1 (hunk 12 above)
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/tools/scripts/GenerateResourceStringsSource.ps1:4]: Added license information as suggessted.

File: src/Directory.Build.props (hunk 9 above)
Comment (DavoudEshtehari) [reply src/Directory.Build.props]: it's still there.

File: src/Directory.Build.props (hunk 10 above)
Comment (DavoudEshtehari) [reply src/Directory.Build.props]: it's still there.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+            // Restore saved culture if necessary
+            if (Thread.CurrentThread.CurrentCulture != savedCulture)
+                Thread.CurrentThread.CurrentCulture = savedCulture;
+            if (Thread.CurrentThread.CurrentUICulture != savedUICulture)
+                Thread.CurrentThread.CurrentUICulture = savedUICulture;
+
+            return localized;
+        }
+
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs:80]:
```suggestion
```

File: src/Directory.Build.props (hunk 10 above)
Comment (arellegue) [reply src/Directory.Build.props]: It was lining up in the editor. But upon close inspection, they were indented using tabs. The tabs were replaced with spaces now.  Since the new lines were copied and pasted, it seemed the editor have replaced spaces with tabs automatically when pasted.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs (hunk 14 above)
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/FunctionalTests/LocalizationTest.cs:80]: Blank line was removed.

File: src/Directory.Build.props (hunk 9 above)
Comment (arellegue) [reply src/Directory.Build.props]: There were tabs instead of spaces.

**Review by DavoudEshtehari — APPROVED** [review]
